Feature/rdkb code porting

morantr opened the pull request:
Port the latest changes from RDKB.

tomereli reviewed (dismissed):
Comments inline

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [port rdkb chnages to prplMesh](https://trello.com/c/i3wVVvpN/25-port-rdkb-chnages-to-prplmesh)

arnout commented on CMakeLists.txt:
```diff
@@ -60,6 +60,7 @@ set(BEEROCKS_BH_WIRE_IFACE "eth1")
 set(BEEROCKS_STOP_ON_FAILURE_ATTEMPTS 1)
 set(BEEROCKS_NOTIFY 0)
 set(BEEROCKS_REPEATER_MODE 0)
+set(BEEROCKS_LOG_PATH "${TMP_PATH}/logs")
```
I want to evolve to a situation where TARGET_PLATFORM is no longer used, and instead we have cache variables that are set by the platform integration (i.e. openwrt Makefile, yocto recipe, or toolchainfile). That is IMO a cleaner way to manage things. So, I would make this a CACHE variable.

Nitpick, because it can easily be done together with moving all the other variations to cache variables.

morantr replied:
will be done as part of a different task.

arnout commented on agent/src/beerocks/monitor/beerocks_monitor_main.cpp:
```diff
@@ -165,8 +165,8 @@ int main(int argc, char *argv[])
-    int fd_log_file_std = beerocks::os_utils::redirect_console_std(beerocks_slave_conf.sLog.path +
-                                                                   base_monitor_name + "_std.log");
+    int fd_log_file_std = beerocks::os_utils::redirect_console_std(
+        beerocks_slave_conf.sLog.path + base_monitor_name + "_std.txt.0");
```
Why the change from `.log` to `.txt.0`?

Since this is such a big (but largely mechanical) change, it should be in a separate commit.

morantr replied:
I really really don't know why they did it. There was no explanation for it in the code I ported.
@tomereli can do you know the reason.

vitalybu replied:
I think that in RDKB there's some sort of centralized logging cleanup service that expects any log files to have specific extension... I think we should keep it for RDKB only.

tomereli replied:
I have no idea. I remember Yaniv did some hacks for rdkb logging and I don't think we should take them as is. Please send him a mail with all the review questions.

arnout commented on common/beerocks/bcl/source/beerocks_logging.cpp:
```diff
@@ -400,7 +400,7 @@ std::string logging::get_log_filename()
         return m_log_filename;
     }
-    return m_module_name + ".%datetime{%Y%M%d_%H%m%s}.log";
+    return m_module_name + ".txt.0";
```
Why remove the datetime (and overwrite the old log)?

The commit message says that you do it, but not why. The why is the important part.

morantr replied:
Same answer.
https://github.com/prplfoundation/prplMesh/pull/537#discussion_r355436664

tomereli replied:
Should be in the mail to Yaniv.

arnout commented on common/beerocks/scripts/beerocks_utils.sh:
```diff
@@ -24,8 +24,9 @@ fi
-BEEROCKS_ERRORS_FILE=$BEEROCKS_TMP_PATH/beerocks_errors.log
+BEEROCKS_LOGS_PATH=/rdklogs/logs
+
+BEEROCKS_ERRORS_FILE=$BEEROCKS_LOGS_PATH/beerocks_errors.txt.0
```
I think it would be better to refactor this script with prplmesh_utils.sh and keep only one of them. That is, assuming that this script is actually still used on rdkb.

morantr replied:
Ok, but in a different task.

tomereli replied:
So please open a new task for it.

arnout commented on framework/CMakeLists.txt:
```diff
@@ -42,6 +42,10 @@ option(BUILD_TESTS "build multiap unit tests" OFF)
 set_property(CACHE MSGLIB PROPERTY STRINGS "zmq" "nng" "None")
 
+if (TARGET_PLATFORM STREQUAL "rdkb")
+	set(MSGLIB "None")
```
As already said by @tomereli : NACK on this. Just pass -DMSGLIB=None while building.

morantr replied:
dropped commit.

arnout commented on controller/src/beerocks/master/tasks/rdkb/rdkb_wlan_task.cpp:
```diff
@@ -266,10 +265,11 @@ void rdkb_wlan_task::handle_event(int event_type, void *obj)
             son_actions::disconnect_client(database, cmdu_tx, client_mac, event_obj->bssid,
                                            event_obj->type, event_obj->reason);
+
+            add_pending_events(int(STEERING_CLIENT_DISCONNECT_RESPONSE), event_obj->sd);
```
The commit message for this change was not really understandable for me. Suggested improvement:

```
controller: rdkb_wlan_task: properly respond to STEERING_CLIENT_DISCONNECT

No response was sent to the STEERING_CLIENT_DISCONNECT_REQUEST,
which caused bml_rdkb_steering_client_disconnect to always report failure.

Queue a response after starting the disconnect_client action.

Also modify some prints.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
```

morantr replied:
changed to your suggested commit message.

arnout commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -567,7 +567,7 @@ bool ap_wlan_hal_dwpal::sta_disassoc(int8_t vap_id, const std::string &mac, uint
     const std::string ifname =
         beerocks::utils::get_iface_string_from_iface_vap_ids(m_radio_info.iface_name, vap_id);
-    std::string cmd = "DISASSOCIATE " + ifname + " " + mac + " tx=0";
+    std::string cmd = "DISASSOCIATE " + ifname + " " + mac;
```
Again, I didn't really understand the commit message. Is the problem that tx=0 is not understood by hostapd? And that this is detected by sniffing the wireless and observing that the disassociate frame doesn't come? Or does tx=0 actually mean to disassociate without sending a disassociate frame?

morantr replied:
I don't know exactly. All I know that the station did not get the dissasoc frame when "tx=0" was in the hostapd command.

tomereli replied:
Not knowing is fine, but putting such changes to the code without knowing and understanding the change is not. Please ask Alex, Yaniv, and whoever was involved with the changes in rdkb (hint - they sit next to you) about this and any other question you didn't know the answer to.
Then add a comment in the code and update the commit message.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -140,7 +147,7 @@ class wireless_utils {
     // clang-format off
     static constexpr sPhyRateTableEntry phy_rate_table[PHY_RATE_TABLE_ANT_MODE_MAX][PHY_RATE_TABLE_MCS_MAX] = {
-            // 1X1_SS1_table
+        // 1X1_SS1_table
```
Should have been part of previous commit.

morantr replied:
I have no idea how this was changed. Anyway, it is not a big deal.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -152,8 +159,8 @@ class wireless_utils {
-            },
-            // 2X2_SS1_table
+        },
+        // 2X2_SS1_table
```
Should have been part of previous commit.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -165,8 +172,8 @@ class wireless_utils {
-            },
-            // 2X2_SS2_table
+        },
+        // 2X2_SS2_table
```
Should have been part of previous commit.

arnout commented on common/beerocks/bcl/include/bcl/son/son_wireless_utils.h:
```diff
@@ -181,7 +188,7 @@ class wireless_utils {
     };
 
-    static constexpr sPhyRateBitRateEntry bit_rate_max_table_mbps[BIT_RATE_MAX_TABLE_SIZE] = {
+    static constexpr sPhyRateBitRateEntry bit_rate_max_table_mbps[BIT_RATE_MAX_TABLE_SIZE] = {  
```
This introduces a whitespace error.

morantr replied:
removed spaces.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -32,20 +32,21 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
-                      : sta_capabilities->ht_mcs;
+    int estimated_ul_rssi_lut_delta_min          = 120 * 10;
+    static const sPhyUlParams estimation_minimal = {
+        is_5ghz ? phy_rate_table[0][0].tx_power_5 : phy_rate_table[0][0].tx_power_2_4,
```
Using a function argument (is_5ghz) to initialize a static variable is not a good idea... In other words, it shouldn't be static.

morantr replied:
done

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -32,20 +32,21 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+    static const int max_mcs =
+        (is_5ghz && (sta_capabilities->wifi_standard & int(beerocks::STANDARD_AC)))
+            ? sta_capabilities->vht_mcs
+            : sta_capabilities->ht_mcs;
```
Again, these can't be static.

morantr replied:
done

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)max_bw);
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
```
This change in the debug output is a bit annoying because it screws up the diff, which complicates review (and also later understanding of this change, in case it needs to be revisited). However, I'm not so sure if it's a useful change.

- If you want to see the function, just set the format to `"%datetime %level [%logger] %func | %msg"` in the logging config.
- I'm not sure what the value is of the additional pipes; it's not really a table, is it? And if it is, it anyway won't be aligned because no field width is defined.

(nitpick)

morantr replied:
Removed the __FUNCTION__.
I agree about the pipes. But the guy who did it said that it looks better in the log file.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+
+    // If station phyrate value is above table's maximum, return maximal estimation
+    if (sta_phy_tx_rate_100kb >
+        phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].gi_short_rate) {
```
It may be appropriate to put this in a variable `max_rate`. That also removes the need for the comment above (it becomes self-explanatory).

Same for the minimum, above.

morantr replied:
I removed the minimal_estimation as you suggested. So I'm leaving it as is.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
-                    (sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate)) {
+                auto gi_long_rate  = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate;
+                auto gi_short_rate = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate;
```
To make the code more readable, it may be more useful to instead define `auto values = phy_rate_table[ant_mode][mcs].bw_values[bw];`. That makes all of the code below much shorter.

morantr replied:
how it will make it shorter? We will have to write `value.gi_long_rate` and  `value.gi_short_rate` instead of `gi_long_rate` and `gi_short_rate`.
Anyway, it doesn't seem critical. I leaving it as is.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+                        estimation_normal.tx_power =
+                            is_5ghz ? phy_rate_table[ant_mode][mcs].tx_power_5
+                                    : phy_rate_table[ant_mode][mcs].tx_power_2_4;
+                        estimation_normal.rssi = int(ceil(estimated_ul_rssi_lut / 10.0));
```
Why ceil instead of approximation rounding like it was before?

morantr replied:
It is the same result. I found it stupid to add 0.5 and then cast it to `int` so the number will be rounded up.

arnout replied:
The original code was `int(float(est) / 10.0 + 0.5)`. Sinds int() rounds down, this rounds down if the last decimal is smaller than 5 and up if it is larger than 5. The new code is `int(ceil(est / 10.0))` which always rounds up.

See this test code:

```
#include <iostream>
#include <cmath>

int main() {
  int est = 14;
  std::cout << "div: " << (float(est) / 10.0) << " int: " << int(float(est)/10.0 + 0.5) << std::endl;
  std::cout << "div: " << (est / 10.0) << " ceil: " << ceil(est/10.0) << " int: " << int(ceil(est/10.0)) << std::endl;
}
```

output:

```
div: 1.4 int: 1
div: 1.4 ceil: 2 int: 2
```

morantr replied:
I thought it will round to the nearest integer value.
Anyway, according to the comments. the one who coded it meant to round it up. So in that case, it fixes it.

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+    // If station phyrate value is below table's minimum, return minimal estimation
+    if (sta_phy_tx_rate_100kb < phy_rate_table[0][0].bw_values[0].gi_long_rate) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (STA phyrate is too low)";
+        return estimation_minimal;
```
Since this is the only place where estimation_minimal is used, I don't think it's worth defining a variable for it. Or at least, move it here. But actually, I'd just set the three fields of estimation_normal here. And then the rename of estimation_normal to estimation is unnecessary as well.

morantr replied:
done

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
+        return estimation_invalid;
```
I don't really think adding the estimation_invalid variable adds a lot of value. The way it was done originally works just as well and is equally clear.

morantr replied:
done

arnout commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -912,6 +913,7 @@ void optimal_path_task::work()
             auto hostap_sibling = it.second;
 
             int hostap_channel    = database.get_node_channel(hostap);
+            skip_estimation       = false; // initialise for each HostAP candidate
```
Then it should also be declared in this scope. So `bool skip_estimation = false;` here.

morantr replied:
done

arnout reviewed (changes requested):
Reviewed up to and including "controller: Add/Modify debug logs and add comments to optimal_path_task".

Most of the commits have excellent commit messages, and also the way it's split up is generally very clean. Kudos!

There's one thing that I believe is an actual bug: making some of the local variables static while they are calculated based on parameters. The static means that they are only calculated the first time.

morantr committed:
Move log files location on RDKB to "/rdklogs/logs/"

Change the location where the log files are written to /rdklogs/logs
in order to be included in the archive that is pushed to the cloud.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Disable arp monitor

The arp_monitor is needed only to discover Ethernet clients.
Disable arp monitor since it observed to cause a high CPU cycle on the agent.
This is a bandaid solution and a more proper solution needs to be found.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: rdkb_wlan_task: properly respond to STEERING_CLIENT_DISCONNECT

No response was sent to the STEERING_CLIENT_DISCONNECT_REQUEST,
which caused bml_rdkb_steering_client_disconnect to always report failure.

Queue a response after starting the disconnect_client action.

Also modify some prints.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
bwl: Fix bug which "disassociate" frames weren't getting captured

Disassociate frames weren't captured when the AP disconnects a client.
Remove 'tx=0' from DISASSOCIATE commands sent to hostapd.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: bwl: dwpal: use C++ initialisation instead of memset

Using memset to initialize a non-POD object is not a good idea.
This was found with Klocwork "memset is applied to a non-POD object".
To solve this we could cast "msg->params.capabilities" to char* but
instead, solve it in the CPP way by setting struct to zero.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller/agent: Port intel patented code

Port code which was wrapped around with #ifdef BEEROCKS_PATENT.

There are two patents which are revealed in this commit:
1. Measuring and calculate AP load and system flow of load
balancing algorithm.
2. Candidate APs picking code for Unconnected measurement
algorithm.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Bugfix -  bwl sockets stay in "/tmp" after the process is dead

Fix bug which causes to  BWL sockets to stay in "/tmp" after the process
has been restarted.
When Hostapd process goes down, the Monitor doesn’t have enough time
to close all its wpa_ctrl sockets.
Solve this by increasing the delay time before Monitor restart, from
2 to 15 seconds top.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Fix Phyrate tables

The clang-format has messed up the Phyrate table.
Fix it and disable the clang on those tables.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add logic and refactor estimate_ul_params function

There was a problem with estimate_ul_params function which often failed
the estimation on the given PhyRate.
Add logics that lookup to values between gi_long_rate to gi_short_rate.
instead of a specific rate.
Also, if the rate is not found between these values, look for it between
the current gi_long_rate to the gi_short_rate of the [mcs-1] which can
be taken since the ranges of the rates are not sequential across the
the table between MCS levels.

Change all rounding to round up instead of round nearest. That is what
was actually intended.

Add also a check if the given PhyRate is above the max rate of the
table, which in that case returning the highest estimation.

Add estimation status to the function return value.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Change calculation of pathloss on estimate_dl_rssi function

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Refactor "which_freq/subband" and "is_high/low_subband" function

Add the use of macros definitions instead of hardcoded numbers.
Use existing functions instead of reimplementing logic.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Fix typo "channnel" to "channel"

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Add/Modify prints on son_wireless_utils

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add skip estimation flag to optimal_path

Add skip estimation flag to optimal_path to prevent re-estimation
which will give the task data it already has.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add handling to estimate_ul_params function failure

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Change optimal_path chosen AP

Change optimal_path chosen AP on signal strength mode, when both bands
are below the cutoff, to always choose the 2.4G AP.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add/Modify debug logs and add comments to optimal_path_task

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

arnout commented on common/beerocks/bcl/source/son/son_wireless_utils.cpp:
```diff
@@ -56,22 +50,27 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+        estimation.tx_power = int(ceil(phy_rate_table[0][0].bw_values[0].rssi / 10.0));
+        estimation.status   = ESTIMATION_FAILURE_BELOW_RANGE;
+
+        return estimation;
```
Yes, I like this a lot better.

Well, except that rssi and tx_power are swapped :-) Not so good...

vitalybu reviewed (commented):
That's not correct. The ARP monitor is used for discovering Ethernet clients (IP address is discovered by the DHCP monitor). AFAIK EasyMesh doesn't handle Ethernet clients, but we already have this built-in...

vitalybu commented on agent/config/beerocks_agent.conf.in:
```diff
@@ -13,7 +13,7 @@ vendor=Intel
 ucc_listener_port=8002 # 0 - disabled
 ucc_listener_slave_hostap_iface=wlan0
-enable_arp_monitor=1
+enable_arp_monitor=0
```
That's not correct. The ARP monitor is used for discovering Ethernet clients (IP address is discovered by the DHCP monitor). AFAIK EasyMesh doesn't handle Ethernet clients, but we already have this built-in...

morantr replied:
In the commit message, I wrote that it observed to cause high CPU cycles and that this is a bandaid solution and better solution must be found. I also added a fixup that disable it only on RDKB.

vitalybu commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -185,6 +185,152 @@ void monitor_stats::process()
+    radio_stats.channel_load_others =
+        radio_stats.channel_load_tot_curr -
+        (radio_stats.client_tx_load_tot_curr + radio_stats.client_rx_load_tot_curr);
+    //LOG(DEBUG) << "STATS_MEASUREMENT_vap: "<< vap_node;
```
Remove commented prints

morantr replied:
done

vitalybu commented on common/beerocks/bcl/source/beerocks_os_utils.cpp:
```diff
@@ -86,10 +87,15 @@ void os_utils::kill_pid(std::string path, std::string file_name)
+        LOG(DEBUG) << __FUNCTION__ << " SIGTERM pid=" << pid_out << std::endl;
         kill(pid_out, SIGTERM);
-        sleep(2); //TODO: find alternative solution
+        auto timeout = std::chrono::steady_clock::now() + std::chrono::seconds(15);
```
Timeout value should be exported as an argument.
But overall, busy-wait is never a good idea.
Maybe sleep for short intervals inside the while loop?
(NOTE: `waitpid()` doesn't support timeouts [without ugly alarm/select workaround])

morantr commented:
> So move it to the BPL and let each platform dB decide

Not as part of this PR. There is already a task for moving configuration to the database.

arnout reviewed (approved):
Except for the rename of the log file, all review feedback was taken into account. So I rebased on master and split the log file commit. Only the specific directory on rdkb is retained; the file renaming is on a separate commit that is pushed to branch dev/change-log-file-names.
